src/main/java/core/characters/CharacterManager.java

```diff
@@ -1,80 +1,145 @@
 package core.characters;
 
+import java.lang.reflect.Array;
 import java.util.ArrayList;
-
-// Need to import InputManager and ControlState
+import java.util.HashMap;
+import java.util.Objects;
 
 public class CharacterManager {
 
     /*
      * Use case class that handles updating instances of GameCharacter based on inputs from the InputHandler
      * @param character: An instance of GameCharacter being updated
+     * @param type: Either Player or AI so the input handler can distinguish between usages
      * TODO: Change item to type Item when implemented
-     * TODO: Import InputManager and ControlState
      * */
 
-    private GameCharacter character;
+    public HashMap<Integer, GameCharacter> characterEntities;
+    private ArrayList<Integer> keylist;
+
+    public CharacterManager() {
+        this.characterEntities = new HashMap<Integer, GameCharacter>();
+        this.keylist = new ArrayList<Integer>();
+    }
 
-    public CharacterManager(GameCharacter character) {
-        this.character = character;
+    public void addCharacter(GameCharacter character) {
+        /*
+        * Player will always be hashed as 0
+        * Defenders will be hashed as even ints
+        * Enemies will be hashed as odd ints
+        * TODO: Refactor
+        * */
+        int maxElement = this.keylist.get(this.keylist.size() - 1);
+
+        if (Objects.equals(character.team, "player")) {
+            this.characterEntities.put(0, character);
+            this.keylist.add(0);
+        } else if (Objects.equals(character.team, "defender")) {
+            if (maxElement % 2 == 0) {
+                this.characterEntities.put(maxElement + 2, character);
+                this.keylist.add(maxElement + 2);
+            } else {
+                this.characterEntities.put(maxElement + 1, character);
+                this.keylist.add(maxElement + 1);
+            }
+        } else {
+            if (maxElement % 2 == 0) {
+                this.characterEntities.put(maxElement + 1, character);
+                this.keylist.add(maxElement + 1);
+            } else {
+                this.characterEntities.put(maxElement + 2, character);
+                this.keylist.add(maxElement + 2);
+            }
+        }
```

Comment:
it's very hard to tell what this does, like sometimes we add 2 sometimes 1 to keylist, but the info there is already based on the character team so why not just use that when necessary?

storing UID as a final variable in the GameCharacters would make it so that we don't need a separate list of all the IDs

Reply:
also if we want to make a single change to this we have to change all the +2 or %2 all over this code